pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -320,12 +321,45 @@ class Utils {
                     tagsAction.addTag(tagName, tagValue)
                     currentNode.save()
                 }
+            }
+        }
+    }
+
+    static void markStartAndEndNodesInStageAsNotExecuted(String stageName, FlowExecution execution = null) {
+        if (execution == null) {
+            CpsThread thread = CpsThread.current()
+            execution = thread.execution
+        }
+
+        List<BlockStartNode> nodes = []
+
+        ForkScanner scanner = new ForkScanner()
+
+        FlowNode stage = scanner.findFirstMatch(execution.currentHeads, null, isStageWithOptionalName(stageName))
 
-                if (isRestart) {
-                    currentNode.addAction(new NotExecutedNodeAction())
+        if (stage != null && stage instanceof BlockStartNode) {
+            nodes.add(stage)
+
+            // Additional check needed to get the possible enclosing parallel branch for a nested stage.
+            Filterator<FlowNode> filtered = FlowScanningUtils.fetchEnclosingBlocks(stage)
+                .filter(isParallelBranchFlowNode(stageName))
+
+            filtered.each { f ->
+                if (f != null && f instanceof BlockStartNode) {
+                    nodes.add(f)
                 }
             }
         }
+
+        DepthFirstScanner depthFirstScanner = new DepthFirstScanner()
+
+        nodes.each { n ->
+            n.addAction(new NotExecutedNodeAction())
+            FlowNode endNode =  depthFirstScanner.findFirstMatch(execution.currentHeads, null, endNodeForStage(n))
```

Comment:
@svanoort - I'm particularly interested to hear whether there's a better way to do this (get the end node for a start node), so comments more than welcome. =)